src/clients/java/src/main/java/com/tigerbeetle/AsyncRequest.java

```diff
@@ -81,7 +85,7 @@ protected void setResult(final TResponse result) {
 
         // To prevent the completion to run in the callback thread
         // we must call "completeAsync" instead of "complete".
-        future.completeAsync(() -> result);
+        future.completeAsync(() -> result, executor != null ? executor : future.defaultExecutor());
```

Comment:
> Returns the default Executor used for async methods that do not specify an Executor. This class uses the [ForkJoinPool.commonPool()](https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/concurrent/ForkJoinPool.html#commonPool()) if it supports more than one parallel thread, or else an Executor using one thread per async task. This method may be overridden in subclasses to return an Executor that provides at least one independent thread.

So looks like this code will still trip thread-per-complete logic.


This makes me think:

_We_ create this future object, so we don’t actually make a meaningful use of virtually dispatch instance method to get the executor —- we know what the executor is going to be from the start.

*And* it looks like we actually want a different logic here —- if the user haven’t supplied the executor, we want to just use the common pool, even if it is single threaded (because we know we are a different thread. My *guess* is that JDK needs that hack to prevent submitting a future from a thread to the same thread).

So, what if we just *require* an executor at this level, and explicitly pass commonPoll at the API layer? That way, we’ll also avoid duplication at the API level, as methods wihout executor will delegate to methods with one.

Reply:
>// To prevent the completion to run in the callback thread
>// we must call "completeAsync" instead of "complete"

Actually, this comment was wrong! We already have the result from the native thread, we don't need to defer it to an async executor. In fact, we already do the same with `completeExceptionally`.
This way, the API stays simple and the user can provide the executor for the completion (if they want to use `thenApply` and friends).

Reply:
I can even remember when/why I wrote this comment, it was shortly after I found an issue in the C# client, where the async continuation would run on the native thread _if_ not stated the opposite:

https://github.com/tigerbeetle/tigerbeetle/blob/280faa17d491a01812321fd6571b7f2f0a03ec64/src/clients/dotnet/TigerBeetle/Request.cs#L145-L146

For some reason, I mixed both languages and applied the same "fix" for Java 🤦‍♂️